src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/ProjectDependencyHelper.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Runtime.Serialization;
+using Microsoft.CodeAnalysis.LanguageServer.LanguageServer;
+using Microsoft.CodeAnalysis.MSBuild;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.Extensions.Logging;
+using NuGet.ProjectModel;
+using NuGet.Versioning;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.HostWorkspace;
+internal static class ProjectDependencyHelper
+{
+    private const string UnresolvedProjectDependenciesName = "workspace/_roslyn_unresolvedProjectDependencies";
+
+    internal static bool HasUnresolvedDependencies(ProjectFileInfo projectFileInfo, ILogger logger)
+    {
+        var projectAssetsPath = projectFileInfo.ProjectAssetsFilePath;
+        if (!File.Exists(projectAssetsPath))
+        {
+            // If the file doesn't exist then all package references are unresolved.
+            logger.LogWarning(string.Format(LanguageServerResources.Project_0_has_unresolved_dependencies, projectFileInfo.FilePath));
+            return true;
+        }
+
+        if (projectFileInfo.PackageReferences.IsEmpty)
+        {
+            // If there are no package references then there are no unresolved dependencies.
```

Comment:
What about stuff from the SDK itself? I admit I don't know the details here but I'm just asking.

Reply:
Looks like there is some information in the assets file on the target frameworks that change if you modify TFMs.  Going to take a look to see if its as simple as checking if the TFMs in the project match the TFMs in the lock file.

Reply:
> Missing stuff from the SDK should be covered under the presence of the project.assets.json.

I wasn't sure about the case where we've seen during dogfooding where you upgrade to some SDK version and the various implicit packs aren't actually in feeds yet.

Reply:
I added some code to look at target framework changes as well.  However I don't think it will be easy to detect SDK version changes - we'd have to track and listen to changes in the global.json or look at which SDK the build host used to load the project (if you manually upgrade without having a global.json).  I don't think that is necessarily worth doing.